### PR TITLE
Vm parser

### PR DESCRIPTION
**Pull request by elevin72**
Opened for the sake of opening. Definitely still work to be done here before merging

### CODE REVIEW

File: lib/vmtranslator/translator.go [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+		return pushPopToHack(cmd), nil
+	case C_ARITHMETIC:
+		return arithmeticToHack(cmd), nil
+	default:
+		//ERROR
+		return "", err
+	}
+}
+
+func pushPopToHack(command *Command) (hack string) {
```
Comment (elevin72) [inline lib/vmtranslator/translator.go:35]: Can this be made 2 separate functions for push and pop?

File: lib/vmtranslator/translator.go [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+		hack = "@THAT\nAD=M\n"
+	case "temp":
+		hack = "@5\nD=A\n"
+	case "pointer":
+		hack = "@THIS\nD=A\n"
+	case "static":
+		return "@" + filename + "." + command.arg2 + "\nD=A\n"
+	}
+
+	if arg2 := command.arg2; arg2 == "0" {
```
Comment (elevin72) [inline lib/vmtranslator/translator.go:79]: hmmm

File: lib/vmtranslator/translator.go [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+	} else if arg2 == "1" {
+		hack += "AD=A+1\n"
+	} else {
+		hack += "@" + arg2 + "\nAD=D+A\n"
+	}
+
+	return hack
+}
+
+var jmpLabel int = 0
```
Comment (elevin72) [inline lib/vmtranslator/translator.go]: Maybe declare all constants in one place/block at the top
```suggestion
const jmpLabel int = 0
```

File: lib/vmtranslator/translator.go [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+		hack += "@SP\nA=M-1\nM=D|M\n"
+	case "not":
+		hack = "@SP\nA=M-1\nM=!M\n"
+	}
+
+	return hack
+}
+
+// popToD Place M[--SP] in D
+func popToD() string {
```
Comment (elevin72) [inline lib/vmtranslator/translator.go]: make const string

File: lib/vmtranslator/translator.go [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+	return "@SP\nAM=M-1\nD=M\n"
+}
+
+// pushFromD Place D in M[SP++]
+func pushFromD() string {
```
Comment (elevin72) [inline lib/vmtranslator/translator.go]: ditto

File: lib/vmtranslator/translator.go [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+	AC_RELOP
+	AC_VAR
+	AC_JUMP
+)
+
+type AsmCommand_t struct{}
+
+type AsmCommand interface{}
+
+func TranslateCommand(cmd *Command) (string, error) {
```
Comment (elevin72) [inline lib/vmtranslator/translator.go]: I agree. No need for error handling at this stage. Or if there is an error then log and panic and die

File: lib/vmtranslator/translator.go [hunk 7]
```diff
@@ -0,0 +1,141 @@
+package vmtranslator
+
+import (
+	"log"
+	"strconv"
+	"strings"
+)
+
+type AC_TYPE int
```
Comment (elevin72) [inline lib/vmtranslator/translator.go]: whats with all these types? are they used for something?

File: lib/vmtranslator/translator.go [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+	case "this":
+		hack = "@THIS\nAD=M\n"
+	case "that":
+		hack = "@THAT\nAD=M\n"
+	case "temp":
+		hack = "@5\nD=A\n"
+	case "pointer":
+		hack = "@THIS\nD=A\n"
+	case "static":
+		return "@" + filename + "." + command.arg2 + "\nD=A\n"
```
Comment (elevin72) [inline lib/vmtranslator/translator.go]: oooh we do need the filename here. hrpmh

File: lib/vmtranslator/parser.go [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+
+type Command struct {
+	cmdType C_TYPE
+	arg1    string
+	arg2    string
+}
+
+func parseCommand(s string) (*Command, error) {
+
+    s = RE_REMOVE_COMMENTS.ReplaceAllLiteralString(s, "")
```
Comment (eliezerjacobs) [inline lib/vmtranslator/parser.go]: will this handle full comment lines correctly?

File: lib/vmtranslator/translator.go (hunk 1 above)
Comment (eliezerjacobs) [reply lib/vmtranslator/translator.go:35]: looks like it. might be cleaner to keep them together cuz they're pretty related to eachother

File: lib/vmtranslator/translator.go (hunk 2 above)
Comment (eliezerjacobs) [reply lib/vmtranslator/translator.go:79]: ?

File: lib/vmtranslator/translator.go (hunk 3 above)
Comment (eliezerjacobs) [reply lib/vmtranslator/translator.go]: i dont think this will work with line 112

File: lib/vmtranslator/translator.go (hunk 7 above)
Comment (ronibuchine) [reply lib/vmtranslator/translator.go]: no ill get rid of them

File: lib/vmtranslator/translator.go (hunk 6 above)
Comment (ronibuchine) [reply lib/vmtranslator/translator.go]: we'll log the error with log.fatal and a helpful error message for us to debug whatever went wrong. no buffer for backend needed

File: lib/vmtranslator/vmtranslator.go [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+
+		parsedCommands := ParseFile(input)
+
+        hack := "// Code Generated from " + input.Name() + "\nPowered by GO (TM)\n"
+		for _, command := range parsedCommands {
+			hackCommand, err := TranslateCommand(command)
+			if err != nil {
+				log.Fatal(err)
+			}
+            hack += hackCommand
```
Comment (eliezerjacobs) [inline lib/vmtranslator/vmtranslator.go]: hack += TranslateCommand(command) ?
Also, there should be an infinite loop added to the end of the program (hack += "@INFLOOP\n(INFLOOP)\n0;JMP\n")

File: lib/vmtranslator/translator.go (hunk 3 above)
Comment (elevin72) [reply lib/vmtranslator/translator.go]: ah you want it to be static? A unique value every time is is accessed throughout runtime?
statics look awkward...
https://www.lucasepe.it/posts/go-closure-vs-static-variables/

File: lib/vmtranslator/vmtranslator.go (hunk 10 above)
Comment (elevin72) [reply lib/vmtranslator/vmtranslator.go]: Why?

**Pull request merged**